README.md
# Scala Exercises - ScalaCheck library
------------------------

This repository hosts a content library for the [Scala Exercises](https://www.scala-exercises.org/) platform, including interactive exercises related to the [ScalaCheck](https://github.com/rickynils/scalacheck) testing library.

## About Scala exercises

"Scala Exercises" brings exercises for the Stdlib, Cats, Shapeless and many other great libraries for Scala to your browser. Offering hundreds of solvable exercises organized into several categories covering the basics of the Scala language and it's most important libraries.

Scala Exercises is available at [scala-exercises.org](https://scala-exercises.org).

## Contributing

Contributions welcome! Please join our [Gitter channel](https://gitter.im/scala-exercises/scala-exercises)
to get involved, or visit our [GitHub site](https://github.com/scala-exercises).

## License

Copyright (C) 2015-2016 47 Degrees, LLC.
Reactive, scalable software solutions.
http://47deg.com
hello@47deg.com

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

CHANGELOG.md
# Changelog

## [v0.6.6](https://github.com/scala-exercises/exercises-scalacheck/tree/v0.6.6) (2020-08-10)

[Full Changelog](https://github.com/scala-exercises/exercises-scalacheck/compare/v0.6.5...v0.6.6)

📈 **Dependency updates**

- Update definitions, exercise-compiler, ... to 0.6.4 [\#82](https://github.com/scala-exercises/exercises-scalacheck/pull/82) ([47erbot](https://github.com/47erbot))

## [v0.6.5](https://github.com/scala-exercises/exercises-scalacheck/tree/v0.6.5) (2020-08-10)

[Full Changelog](https://github.com/scala-exercises/exercises-scalacheck/compare/v0.6.4...v0.6.5)

📈 **Dependency updates**

- Update sbt-mdoc to 2.2.4 [\#81](https://github.com/scala-exercises/exercises-scalacheck/pull/81) ([47erbot](https://github.com/47erbot))
- Update scalacheck-1-14 to 3.2.1.0 [\#80](https://github.com/scala-exercises/exercises-scalacheck/pull/80) ([47erbot](https://github.com/47erbot))
- Update scalatest to 3.2.1 [\#79](https://github.com/scala-exercises/exercises-scalacheck/pull/79) ([47erbot](https://github.com/47erbot))
- Update sbt-scalafmt to 2.4.2 [\#78](https://github.com/scala-exercises/exercises-scalacheck/pull/78) ([47erbot](https://github.com/47erbot))
- Update scalafmt-core to 2.6.4 [\#77](https://github.com/scala-exercises/exercises-scalacheck/pull/77) ([47erbot](https://github.com/47erbot))
- Update scalafmt-core to 2.6.3 [\#76](https://github.com/scala-exercises/exercises-scalacheck/pull/76) ([scala-steward](https://github.com/scala-steward))
- Update scalafmt-core to 2.6.2 [\#75](https://github.com/scala-exercises/exercises-scalacheck/pull/75) ([scala-steward](https://github.com/scala-steward))

## [v0.6.4](https://github.com/scala-exercises/exercises-scalacheck/tree/v0.6.4) (2020-06-30)

[Full Changelog](https://github.com/scala-exercises/exercises-scalacheck/compare/v0.6.3...v0.6.4)

## [v0.6.3](https://github.com/scala-exercises/exercises-scalacheck/tree/v0.6.3) (2020-06-29)

[Full Changelog](https://github.com/scala-exercises/exercises-scalacheck/compare/v0.6.2...v0.6.3)

📈 **Dependency updates**

- Update sbt to 1.3.13 [\#71](https://github.com/scala-exercises/exercises-scalacheck/pull/71) ([scala-steward](https://github.com/scala-steward))
- Update sbt-mdoc to 2.2.3 [\#70](https://github.com/scala-exercises/exercises-scalacheck/pull/70) ([scala-steward](https://github.com/scala-steward))
- Update scalacheck-1-14 to 3.2.0.0 [\#69](https://github.com/scala-exercises/exercises-scalacheck/pull/69) ([scala-steward](https://github.com/scala-steward))
- Update scalatest to 3.2.0 [\#68](https://github.com/scala-exercises/exercises-scalacheck/pull/68) ([scala-steward](https://github.com/scala-steward))
- Update scalafmt-core to 2.6.1 [\#67](https://github.com/scala-exercises/exercises-scalacheck/pull/67) ([scala-steward](https://github.com/scala-steward))
- Update sbt-mdoc to 2.2.2 [\#66](https://github.com/scala-exercises/exercises-scalacheck/pull/66) ([scala-steward](https://github.com/scala-steward))
- Update sbt to 1.3.11 [\#63](https://github.com/scala-exercises/exercises-scalacheck/pull/63) ([scala-steward](https://github.com/scala-steward))
- Update scalatest to 3.1.2 [\#62](https://github.com/scala-exercises/exercises-scalacheck/pull/62) ([scala-steward](https://github.com/scala-steward))

**Merged pull requests:**

- Update definitions, exercise-compiler, ... to 0.6.3 [\#72](https://github.com/scala-exercises/exercises-scalacheck/pull/72) ([scala-steward](https://github.com/scala-steward))
- Prepare repository for next  release and SBT build improvements [\#65](https://github.com/scala-exercises/exercises-scalacheck/pull/65) ([juanpedromoreno](https://github.com/juanpedromoreno))
- Add note about Checkers trait [\#58](https://github.com/scala-exercises/exercises-scalacheck/pull/58) ([aleksandr-vin](https://github.com/aleksandr-vin))

## [v0.6.2](https://github.com/scala-exercises/exercises-scalacheck/tree/v0.6.2) (2020-04-27)

[Full Changelog](https://github.com/scala-exercises/exercises-scalacheck/compare/v0.6.1...v0.6.2)

## [v0.6.1](https://github.com/scala-exercises/exercises-scalacheck/tree/v0.6.1) (2020-04-27)

[Full Changelog](https://github.com/scala-exercises/exercises-scalacheck/compare/v0.6.0...v0.6.1)

📈 **Dependency updates**

- Update joda-time to 2.10.6 [\#53](https://github.com/scala-exercises/exercises-scalacheck/pull/53) ([scala-steward](https://github.com/scala-steward))

**Merged pull requests:**

- Compile when Publish [\#55](https://github.com/scala-exercises/exercises-scalacheck/pull/55) ([juanpedromoreno](https://github.com/juanpedromoreno))

## [v0.6.0](https://github.com/scala-exercises/exercises-scalacheck/tree/v0.6.0) (2020-04-24)

[Full Changelog](https://github.com/scala-exercises/exercises-scalacheck/compare/v0.4.0...v0.6.0)

**Merged pull requests:**

- Updates Project [\#50](https://github.com/scala-exercises/exercises-scalacheck/pull/50) ([juanpedromoreno](https://github.com/juanpedromoreno))
- Update scalacheck-toolbox-datetime to 0.3.5 [\#48](https://github.com/scala-exercises/exercises-scalacheck/pull/48) ([scala-steward](https://github.com/scala-steward))
- Update sbt to 1.3.10 [\#47](https://github.com/scala-exercises/exercises-scalacheck/pull/47) ([scala-steward](https://github.com/scala-steward))
- Update scalacheck-toolbox-datetime to 0.3.4 [\#46](https://github.com/scala-exercises/exercises-scalacheck/pull/46) ([scala-steward](https://github.com/scala-steward))
- Update sbt to 1.3.9 [\#45](https://github.com/scala-exercises/exercises-scalacheck/pull/45) ([scala-steward](https://github.com/scala-steward))
- Update scalacheck-toolbox-datetime to 0.3.3 [\#43](https://github.com/scala-exercises/exercises-scalacheck/pull/43) ([scala-steward](https://github.com/scala-steward))
- Update scalacheck-shapeless\_1.14 to 1.2.5 [\#42](https://github.com/scala-exercises/exercises-scalacheck/pull/42) ([scala-steward](https://github.com/scala-steward))

src/test/scala/scalachecklib/Test.scala
/*
 * scala-exercises - exercises-scalacheck
 * Copyright (C) 2015-2016 47 Degrees, LLC. <http://www.47deg.com>
 */

package scalachecklib

import shapeless._
import shapeless.ops.function._

import cats.implicits._
import org.scalacheck.{Arbitrary, Prop}
import org.scalacheck.Gen
import Prop.forAll

import org.scalatest.exceptions._

import org.scalacheck.Shapeless._

object Test {

  def testSuccess[F, R, L <: HList](method: F, answer: L)(
      implicit A: Arbitrary[L],
      fntop: FnToProduct.Aux[F, L ⇒ R]
  ): Prop = {
    val rightGen = genRightAnswer(answer)
    val rightProp = forAll(rightGen)({ p ⇒
      val result = Either.catchOnly[GeneratorDrivenPropertyCheckFailedException]({
        fntop(method)(p)
      })
      result match {
        case Left(exc) ⇒
          exc.cause match {
            case Some(originalException) ⇒ throw originalException
            case _                       ⇒ false
          }
        case _ ⇒ true
      }
    })

    val wrongGen = genWrongAnswer(answer)
    val wrongProp = forAll(wrongGen)({ p ⇒
      Either.catchNonFatal({ fntop(method)(p) }).isLeft
    })

    Prop.all(rightProp, wrongProp)
  }

  def genRightAnswer[L <: HList](answer: L): Gen[L] =
    Gen.const(answer)

  def genWrongAnswer[L <: HList](l: L)(
      implicit A: Arbitrary[L]
  ): Gen[L] =
    A.arbitrary.suchThat(_ != l)
}


src/test/scala/scalachecklib/ArbitrarySpec.scala
/*
 * Copyright 2016-2020 47 Degrees Open Source <https://www.47deg.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package scalachecklib

import org.scalacheck.ScalacheckShapeless._
import org.scalaexercises.Test
import org.scalatest.funsuite.AnyFunSuite
import org.scalatestplus.scalacheck.Checkers
import shapeless.HNil

class ArbitrarySpec extends AnyFunSuite with Checkers {

  test("implicit arbitrary char") {

    check(
      Test.testSuccess(
        ArbitrarySection.implicitArbitraryChar _,
        Seq('A', 'E', 'I', 'O', 'U') :: HNil
      )
    )
  }

  test("implicit arbitrary case class") {

    check(
      Test.testSuccess(
        ArbitrarySection.implicitArbitraryCaseClass _,
        false :: HNil
      )
    )
  }

  test("arbitrary on gen") {

    check(
      Test.testSuccess(
        ArbitrarySection.useArbitraryOnGen _,
        8 :: HNil
      )
    )
  }

}


src/test/scala/exercises/scalachecklib/Test.scala
package exercises.scalachecklib

import cats.data.Xor

import shapeless._
import shapeless.ops.function._

import org.scalacheck.{Prop, Arbitrary }
import org.scalacheck.Gen
import org.scalacheck.Shapeless._
import org.scalatest.exceptions._

import Prop.forAll

object Test {

  def testSuccess[F, R, L <: HList](method: F, answer: L)(
    implicit
    A:     Arbitrary[L],
    fntop: FnToProduct.Aux[F, L ⇒ R]
  ): Prop = {
    val rightGen = genRightAnswer(answer)
    val rightProp = forAll(rightGen)({ p ⇒

      val result = Xor.catchOnly[GeneratorDrivenPropertyCheckFailedException]({ fntop(method)(p) })
      result match {
        case Xor.Left(exc) ⇒ exc.cause match {
          case Some(originalException) ⇒ throw originalException
          case _                       ⇒ false
        }
        case _ ⇒ true
      }
    })

    val wrongGen = genWrongAnswer(answer)
    val wrongProp = forAll(wrongGen)({ p ⇒
      Xor.catchNonFatal({ fntop(method)(p) }).isLeft
    })

    Prop.all(rightProp, wrongProp)
  }

  def genRightAnswer[L <: HList](answer: L): Gen[L] = {
    Gen.const(answer)
  }

  def genWrongAnswer[L <: HList](l: L)(
    implicit
    A: Arbitrary[L]
  ): Gen[L] = {
    A.arbitrary.suchThat(_ != l)
  }
}



src/main/scala/scalachecklib/ArbitrarySection.scala
/*
 * Copyright 2016-2020 47 Degrees Open Source <https://www.47deg.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package scalachecklib

import org.scalatest.matchers.should.Matchers
import org.scalatestplus.scalacheck.Checkers

/**
 * ==The `arbitrary` Generator==
 *
 * There is a special generator, `org.scalacheck.Arbitrary.arbitrary`, which generates arbitrary values of any
 * supported type.
 *
 * {{{
 * val evenInteger = Arbitrary.arbitrary[Int] suchThat (_ % 2 == 0)
 * val squares = for {
 *   xs <- Arbitrary.arbitrary[List[Int]]
 * } yield xs.map(x => x*x)
 * }}}
 *
 * The `arbitrary` generator is the generator used by ScalaCheck when it generates values for ''property'' parameters.
 * Most of the times, you have to supply the type of the value to `arbitrary`, like above, since Scala often can't
 * infer the type automatically. You can use `arbitrary` for any type that has an implicit `Arbitrary` instance.
 * As mentioned earlier, ScalaCheck has default support for common types, but it is also possible to define your own
 * implicit `Arbitrary` instances for unsupported types. See the following implicit Arbitrary definition for booleans,
 * that comes from the ScalaCheck implementation.
 *
 * {{{
 * implicit lazy val arbBool: Arbitrary[Boolean] = Arbitrary(oneOf(true, false))
 * }}}
 *
 * @param name arbitrary
 */
object ArbitrarySection extends Checkers with Matchers with org.scalaexercises.definitions.Section {

  import GeneratorsHelper._

  /**
   * Let's see an example where we're defining an `implicit` `arbitrary` instance for `Char`
   */
  def implicitArbitraryChar(res0: Seq[Char]) = {

    import org.scalacheck.Arbitrary
    import org.scalacheck.Gen
    import org.scalacheck.Prop.forAll

    implicit lazy val myCharArbitrary = Arbitrary(Gen.oneOf('A', 'E', 'I', 'O', 'U'))

    val validChars: Seq[Char] = res0

    check(forAll { c: Char => validChars.contains(c) })
  }

  /**
   * This becomes more useful when we're dealing with our own data types.
   * We'll use the case class defined in the ''Generators Section'':
   *
   * {{{
   * case class Foo(intValue: Int, charValue: Char)
   * }}}
   *
   * Having an implicit `def` or `val` of our data type in the scope allow us to use the `forAll` method without
   * specifying the ''generator''
   */
  def implicitArbitraryCaseClass(res0: Boolean) = {

    import org.scalacheck.Arbitrary
    import org.scalacheck.Gen
    import org.scalacheck.Prop.forAll

    val fooGen = for {
      intValue  <- Gen.posNum[Int]
      charValue <- Gen.alphaChar
    } yield Foo(intValue, charValue)

    implicit lazy val myFooArbitrary = Arbitrary(fooGen)

    check(forAll { foo: Foo => (foo.intValue < 0) == res0 && !foo.charValue.isDigit })
  }

  /**
   * The `Arbitrary.arbitrary` method also returns a `Gen` object.
   */
  def useArbitraryOnGen(res0: Int) = {

    import org.scalacheck.Arbitrary
    import org.scalacheck.Gen.listOfN
    import org.scalacheck.Prop.forAll

    val genEightBytes = listOfN(8, Arbitrary.arbitrary[Byte])

    check(forAll(genEightBytes)(list => list.size == res0))

  }

}


src/main/scala/scalachecklib/GeneratorsSection.scala
/*
 * Copyright 2016-2020 47 Degrees Open Source <https://www.47deg.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package scalachecklib

import org.scalatest.matchers.should.Matchers
import org.scalatestplus.scalacheck.Checkers

/**
 * Generators are responsible for generating test data in ScalaCheck, and are represented by the `org.scalacheck.Gen`
 * class. ou need to know how to use this class if you want ScalaCheck to generate data of types that are not supported
 * by default, or if you want to use the `forAll` method mentioned above, to state properties about a specific subset of
 * a type. In the `Gen` object, there are several methods for creating new and modifying existing generators.
 * We will show how to use some of them in this section. For a more complete reference of what is available,
 * please see the API scaladoc.
 *
 * A generator can be seen simply as a function that takes some generation parameters, and (maybe) returns a
 * generated value. That is, the type `Gen[T]` may be thought of as a function of type `Gen.Params => Option[T]`.
 * However, the `Gen` class contains additional methods to make it possible to map generators, use them in
 * for-comprehensions and so on. Conceptually, though, you should think of generators simply as functions, and the
 * combinators in the `Gen` object can be used to create or modify the behaviour of such generator functions.
 *
 * @param name generators
 */
object GeneratorsSection
    extends Checkers
    with Matchers
    with org.scalaexercises.definitions.Section {

  import GeneratorsHelper._

  /**
   * Let's see how to create a new generator. The best way to do it is to use the generator combinators that exist
   * in the `org.scalacheck.Gen` module. These can be combined using a for-comprehension. Suppose you need a generator
   * which generates a tuple that contains two random integer values, one of them being at least twice as big as the
   * other. The following definition does this:
   */
  def forComprehension(res0: Boolean) = {

    import org.scalacheck.Gen
    import org.scalacheck.Prop.forAll

    val myGen = for {
      n <- Gen.choose(10, 20)
      m <- Gen.choose(2 * n, 500)
    } yield (n, m)

    check {
      forAll(myGen) {
        case (n, m) => (m >= 2 * n) == res0
      }
    }

  }

  /**
   * You can create generators that pick one value out of a selection of values.
   * The `oneOf` method creates a generator that randomly picks one of its parameters each time it generates a value.
   * Notice that plain values are implicitly converted to generators (which always generate that value) if needed.
   *
   * The following generator generates a vowel:
   */
  def genOf(res0: Seq[Char]) = {

    import org.scalacheck.Gen
    import org.scalacheck.Prop.forAll

    val vowel = Gen.oneOf('A', 'E', 'I', 'O', 'U')

    val validChars: Seq[Char] = res0

    check {
      forAll(vowel)(v => validChars.contains(v))
    }
  }

  /**
   * The distribution is uniform, but if you want to control it you can use the frequency combinator:
   *
   * {{{
   * val vowel = Gen.frequency(
   *   (3, 'A'),
   *   (4, 'E'),
   *   (2, 'I'),
   *   (3, 'O'),
   *   (1, 'U')
   * )
   * }}}
   *
   * Now, the vowel generator will generate ''E:s'' more often than ''U:s''. Roughly, 4/14 of the values generated
   * will be ''E:s'', and 1/14 of them will be ''U:s''.
   *
   * Other methods in the `Gen` API:
   * {{{
   * def alphaChar: Gen[Char]
   *
   * def alphaStr: Gen[String]
   *
   * def posNum[T](implicit n: Numeric[T]): Gen[T]
   *
   * def listOf[T](g: Gen[T]): Gen[List[T]]
   *
   * def listOfN[T](n: Int, g: Gen[T]): Gen[List[T]]
   * }}}
   */
  def genAPI(res0: Boolean, res1: Boolean, res2: Int) = {

    import org.scalacheck.Gen.{alphaChar, listOfN, posNum}
    import org.scalacheck.Prop.forAll

    check {
      forAll(alphaChar)(_.isDigit == res0)
    }

    check {
      forAll(posNum[Int])(n => (n > 0) == res1)
    }

    check {
      forAll(listOfN(10, posNum[Int]))(list => !list.exists(_ < 0) && list.length == res2)
    }
  }

  /**
   * ==Conditional Generators==
   *
   * Conditional generators can be defined using `Gen.suchThat`.
   *
   * Conditional generators works just like conditional properties, in the sense that if the condition is too hard,
   * ScalaCheck might not be able to generate enough values, and it might report a property test as undecided.
   * The `smallEvenInteger` definition is probably OK, since it will only throw away half of the generated numbers,
   * but one has to be careful when using the `suchThat` operator.
   */
  def conditionalOperators(res0: Int) = {

    import org.scalacheck.Gen
    import org.scalacheck.Prop.forAll

    val smallEvenInteger = Gen.choose(0, 200) suchThat (_ % 2 == 0)

    check {
      forAll(smallEvenInteger)(_ % 2 == res0)
    }
  }

  /**
   * ==Case class Generators==
   *
   * On the basis of the above we can create a generator for the following case class:
   *
   * {{{
   * case class Foo(intValue: Int, charValue: Char)
   * }}}
   */
  def caseClassGenerator(res0: Boolean) = {

    import org.scalacheck.Gen
    import org.scalacheck.Prop.forAll

    val fooGen = for {
      intValue  <- Gen.posNum[Int]
      charValue <- Gen.alphaChar
    } yield Foo(intValue, charValue)

    check {
      forAll(fooGen)(foo => foo.intValue > 0 && foo.charValue.isDigit == res0)
    }
  }

  /**
   * ==Sized Generators==
   *
   * When ScalaCheck uses a generator to generate a value, it feeds it with some parameters. One of the parameters
   * the generator is given is a size value, which some generators use to generate their values.
   *
   * If you want to use the size parameter in your own generator, you can use the `Gen.sized` method:
   *
   * {{{
   * def sized[T](f: Int => Gen[T])
   * }}}
   *
   * In this example we're creating a generator that produces two lists of numbers where 1/3 are positive and 2/3 are
   * negative. ''Note: we're also returning the original size to verify the behaviour.''
   */
  def sizedGenerator(res0: Int, res1: Int) = {

    import org.scalacheck.Gen
    import org.scalacheck.Prop.forAll

    val myGen = Gen.sized { size =>
      val positiveNumbers = size / 3
      val negativeNumbers = size * 2 / 3
      for {
        posNumList <- Gen.listOfN(positiveNumbers, Gen.posNum[Int])
        negNumList <- Gen.listOfN(negativeNumbers, Gen.posNum[Int] map (n => -n))
      } yield (size, posNumList, negNumList)
    }

    check {
      forAll(myGen) {
        case (genSize, posN, negN) =>
          posN.length == genSize / res0 && negN.length == genSize * res1 / 3
      }
    }
  }

  /**
   * ==Generating Containers==
   *
   * There is a special generator, `Gen.containerOf`, that generates containers such as lists and arrays.
   * It takes another generator as argument which is responsible for generating the individual items.
   * You can use it in the following way:
   *
   * {{{
   * val genIntList      = Gen.containerOf[List,Int](Gen.oneOf(1, 3, 5))
   *
   * val genStringStream = Gen.containerOf[LazyList,String](Gen.alphaStr)
   *
   * val genBoolArray    = Gen.containerOf[Array,Boolean](true)
   * }}}
   *
   * By default, ScalaCheck supports generation of `List`, `Stream` (Scala 2.10 -
   * 2.12, deprecated in 2.13), `LazyList` (Scala 2.13), `Set`, `Array`, and
   * `ArrayList` (from `java.util`). You can add support for additional containers
   * by adding implicit `Buildable` instances. See `Buildable.scala` for examples.
   *
   * There is also `Gen.nonEmptyContainerOf` for generating non-empty containers, and `Gen.containerOfN` for
   * generating containers of a given size.
   */
  def generatingContainers(res0: List[Int]) = {

    import org.scalacheck.Gen
    import org.scalacheck.Prop.forAll

    val genIntList = Gen.containerOf[List, Int](Gen.oneOf(2, 4, 6))

    val validNumbers: List[Int] = res0

    check {
      forAll(genIntList)(_ forall (elem => validNumbers.contains(elem)))
    }
  }

}


src/main/scala/scalachecklib/PropertiesSection.scala
/*
 * Copyright 2016-2020 47 Degrees Open Source <https://www.47deg.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package scalachecklib

import org.scalatest.matchers.should.Matchers
import org.scalatestplus.scalacheck.Checkers

/**
 * A ''property'' is the testable unit in ScalaCheck, and is represented by the `org.scalacheck.Prop` class.
 * There are several ways to create properties in ScalaCheck, one of them is to use the `org.scalacheck.Prop.forAll`
 * method like in the example below.
 *
 * {{{
 * scala> val propConcatLists = forAll { (l1: List[Int], l2: List[Int]) =>
 *   l1.size + l2.size == (l1 ::: l2).size }
 * }}}
 *
 * We can use the `check` method to test it:
 *
 * {{{
 * scala> propConcatLists.check
 * + OK, passed 100 tests.
 * }}}
 *
 * OK, that seemed alright. Now, we'll define another property.
 *
 * {{{
 * scala> val propSqrt = forAll { (n: Int) => scala.math.sqrt(n*n) == n }
 * }}}
 *
 * And check it:
 *
 * {{{
 * scala> propSqrt.check
 * ! Falsified after 2 passed tests.
 * > ARG_0: -1
 * > ARG_0_ORIGINAL: -488187735
 * }}}
 *
 * Not surprisingly, the property doesn't hold. The argument `-1` falsifies it. You can also see that the argument
 * `-488187735` falsifies the property. That was the first argument ScalaCheck found, and it was then
 * [[https://github.com/rickynils/scalacheck/blob/master/doc/UserGuide.md#test-case-minimisation simplified]] to
 * `-1`.
 *
 * The `forAll` method creates universally quantified properties directly, but it is also
 * possible to create new properties by combining other properties, or to use any of the specialised
 * methods in the `org.scalacheck.Prop` object.
 *
 * @param name properties
 */
object PropertiesSection
    extends Checkers
    with Matchers
    with org.scalaexercises.definitions.Section {

  /**
   * ==Universally quantified properties==
   *
   * As mentioned before, `org.scalacheck.Prop.forAll` creates universally quantified properties.
   * `forAll` takes a function as parameter, and creates a property out of it that can be tested with the `check`
   * method or with Scalatest (using Checkers trait), like in these examples.
   *
   * The function passed to `forAll` should return `Boolean` or another property, and can take parameters of any types,
   * as long as there exist implicit `Arbitrary` instances for the types.
   * By default, ScalaCheck has instances for common types like `Int`, `String`, `List`, etc, but it is also possible
   * to define your own `Arbitrary` instances.
   *
   * For example:
   */
  def universallyQuantifiedPropertiesString(res0: Boolean) = {

    import org.scalacheck.Prop.forAll

    check {
      forAll((s1: String, s2: String) => (s1 + s2).endsWith(s2) == res0)
    }

  }

  /**
   * When you run `check` on the properties, ScalaCheck generates random instances of the function parameters and
   * evaluates the results, reporting any failing cases.
   *
   * You can also give `forAll` a specific data generator. In the following example `smallInteger` defines a generator
   * that generates integers between `0` and `100`, inclusively.
   *
   * This way of using the `forAll` method is good to use when you want to control the data generation by specifying
   * exactly which generator that should be used, and not rely on a default generator for the given type.
   */
  def universallyQuantifiedPropertiesGen(res0: Boolean) = {

    import org.scalacheck.Gen
    import org.scalacheck.Prop.forAll

    val smallInteger = Gen.choose(0, 100)

    check {
      forAll(smallInteger)(n => (n >= 0 && n <= 100) == res0)
    }

  }

  /**
   * ==Conditional properties==
   *
   * Sometimes, a specification takes the form of an implication. In ScalaCheck, you can use the implication
   * operator `==>` to filter the generated values.
   *
   * If the implication operator is given a condition that is hard or impossible to fulfill, ScalaCheck might
   * not find enough passing test cases to state that the property holds. In the following trivial example,
   * all cases where `n` is non-zero will be thrown away:
   *
   * {{{
   * scala> import org.scalacheck.Prop.{forAll, propBoolean}
   * scala> val propTrivial = forAll { n: Int =>
   *      |  (n == 0) ==> n + 10 == 10
   *      | }
   *
   * scala> propTrivial.check
   * ! Gave up after only 4 passed tests. 500 tests were discarded.
   * }}}
   *
   * It is possible to tell ScalaCheck to try harder when it generates test cases, but generally you should
   * try to refactor your property specification instead of generating more test cases, if you get this scenario.
   *
   * Using implications, we realise that a property might not just pass or fail, it could also be undecided if
   * the implication condition doesn't get fulfilled.
   *
   * In this example, ScalaCheck will only care for the cases when `n` is an even number.
   */
  def conditionalProperties(res0: Int) = {

    import org.scalacheck.Prop.{forAll, propBoolean}

    check {
      forAll { n: Int => (n % 2 == 0) ==> (n % 2 == res0) }
    }

  }

  /**
   * ==Combining Properties==
   *
   * A third way of creating properties, is to combine existing properties into new ones.
   *
   * {{{
   * val p1 = forAll(...)
   * val p2 = forAll(...)
   * val p3 = p1 && p2
   * val p4 = p1 || p2
   * val p5 = p1 == p2
   * val p6 = all(p1, p2) // same as p1 && p2
   * val p7 = atLeastOne(p1, p2) // same as p1 || p2
   * }}}
   *
   * Here, `p3` will hold if and only if both `p1` and `p2` hold, `p4` will hold if either `p1` or `p2` holds,
   * and `p5` will hold if `p1` holds exactly when `p2` holds and vice versa.
   */
  def combiningProperties(res0: Boolean, res1: Boolean) = {

    import org.scalacheck.Gen
    import org.scalacheck.Prop.forAll

    val smallInteger = Gen.choose(0, 100)

    check {
      forAll(smallInteger) { n =>
        (n > 100) == res0
      } &&
      forAll(smallInteger)(n => (n >= 0) == res1)
    }

  }

  /**
   * ==Grouping Properties==
   *
   * Often you want to specify several related properties, perhaps for all methods in a class.
   * ScalaCheck provides a simple way of doing this, through the `Properties` trait.
   * Look at the following specifications that define some properties for zero.
   *
   * You can use the check method of the `Properties` class to check all specified properties,
   * just like for simple `Prop` instances. In fact, `Properties` is a subtype of `Prop`,
   * so you can use it just as if it was a single property.
   *
   * That single property holds if and only if all of the contained properties hold.
   */
  def groupingProperties(res0: Int, res1: Int, res2: Int) = {
    import org.scalacheck.{Prop, Properties}

    class ZeroSpecification extends Properties("Zero") {

      import org.scalacheck.Prop.{forAll, propBoolean}

      property("addition property") = forAll { n: Int => (n != 0) ==> (n + res0 == n) }

      property("additive inverse property") = forAll { n: Int => (n != 0) ==> (n + (-n) == res1) }

      property("multiplication property") = forAll { n: Int => (n != 0) ==> (n * res2 == 0) }

    }

    check(Prop.all(new ZeroSpecification().properties.to(List).map(_._2): _*))
  }
}


src/main/scala/scalachecklib/ScalacheckLibrary.scala
/*
 * Copyright 2016-2020 47 Degrees Open Source <https://www.47deg.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package scalachecklib

/**
 * ScalaCheck is a tool for testing Scala and Java programs, based on property specifications and automatic test data generation.
 *
 * @param name scalacheck
 */
object ScalacheckLibrary extends org.scalaexercises.definitions.Library {
  override def owner      = "scala-exercises"
  override def repository = "exercises-scalacheck"

  override def color = Some("#EBC477")

  override def sections =
    List(
      PropertiesSection,
      GeneratorsSection,
      ArbitrarySection,
      ScalacheckDatetimeSection
    )

  override def logoPath = "scalacheck"
}


project/ProjectPlugin.scala
import sbt.Keys._
import sbt._
import com.alejandrohdezma.sbt.github.SbtGithubPlugin

object ProjectPlugin extends AutoPlugin {

  override def trigger: PluginTrigger = allRequirements

  override def requires: Plugins = plugins.JvmPlugin && SbtGithubPlugin

  object autoImport {

    lazy val V = new {
      val jodaTime: String            = "2.10.6"
      val shapeless: String           = "2.3.3"
      val scala: String               = "2.13.2"
      val scalaExercises: String      = "0.6.0"
      val scalacheck: String          = "1.14.3"
      val scalacheckDatetime: String  = "0.3.5"
      val scalacheckShapeless: String = "1.2.5"
      val scalatest: String           = "3.1.2"
      val scalatestplusScheck: String = "3.1.1.1"
    }

    def dep(artifactId: String) = "org.scala-exercises" %% artifactId % V.scalaExercises

    lazy val exercisesSettings = Seq(
      libraryDependencies ++= Seq(
        dep("exercise-compiler"),
        dep("definitions"),
        "joda-time"                  % "joda-time"                    % V.jodaTime,
        "com.47deg"                  %% "scalacheck-toolbox-datetime" % V.scalacheckDatetime,
        "com.chuusai"                %% "shapeless"                   % V.shapeless,
        "org.scalatest"              %% "scalatest"                   % V.scalatest,
        "org.scalacheck"             %% "scalacheck"                  % V.scalacheck,
        "org.scalatestplus"          %% "scalacheck-1-14"             % V.scalatestplusScheck,
        "com.github.alexarchambault" %% "scalacheck-shapeless_1.14"   % V.scalacheckShapeless
      )
    )
  }

  override def projectSettings: Seq[Def.Setting[_]] =
    Seq(
      organization := "org.scala-exercises",
      organizationName := "47 Degrees",
      organizationHomepage := Some(url("https://47deg.com")),
      scalaVersion := autoImport.V.scala,
      resolvers ++= Seq(
        Resolver.sonatypeRepo("snapshots"),
        Resolver.sonatypeRepo("releases")
      )
    )
}
